src/link/output.c

```diff
@@ -273,6 +275,47 @@ static struct SortedSection const **nextSection(struct SortedSection const **s1,
 	return (*s1)->section->org < (*s2)->section->org ? s1 : s2;
 }
 
+// Checks whether a symbol is
+static bool canStartSymName(char c)
+{
+	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_';
+}
+
+static void printSymName(char const *name)
+{
+	for (char const *ptr = name; *ptr != '\0'; ) {
+		char c = *ptr;
+
+		if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')
+		 || c == '_' || c == '@' || c == '#' || c == '$' || c == '.') {
+			fputc(c, symFile);
+		++ptr;
+		} else {
+			// Decode the UTF-8 codepoint; or at least attempt to
+			uint32_t state = 0, codepoint;
+
+			do {
+				decode(&state, &codepoint, *ptr);
+				if (state == 1) {
+					// This sequence was invalid; emit a U+FFFD, and recover
+					codepoint = 0xFFFD;
+					// Skip continuation bytes
+					// A NUL byte does not qualify, so we're good
+					while ((*ptr & 0xC0) == 0x80)
+						++ptr;
```

Comment:
Can this logic be done by calling `decode` instead, to avoid repeating these details of UTF-8 mask values outside the utf8decoder?

Reply:
I've looked at it, but `decode` stays in state `1` once falling into it. As noted under that "Error recovery" link, errors are detected ASAP, so they would tend to produce many `�`s (one per continuation byte, give or take), which would spam output. Hence this heuristic.